persist/fs/commitlog/reader.go

```diff
@@ -110,52 +157,132 @@ func (r *reader) Read() (
 	annotation ts.Annotation,
 	resultErr error,
 ) {
-	entry, err := r.readEntry()
-	if err != nil {
-		resultErr = err
-		return
+	if r.nextIndex == 0 {
+		go r.readLoop()
+		for i, decoderBuf := range r.decoderBufs {
+			go r.decoderLoop(decoderBuf, r.outBufs[i])
+		}
 	}
+	// Data is written into these channels in round-robin fashion, so reading them
+	// one at a time in the same order results in an ordered stream.
+	// TODO: This doesn't eventually deadlock because the reader goroutine will
+	// continue to fill these channels with EOFs until it is closed. We should
+	// probably avoid the deadlock (read from empty channel) scenario in a more
+	// graceful way.
+	rr := <-r.outBufs[r.nextIndex%r.numConc]
+	r.nextIndex++
+	return rr.series, rr.datapoint, rr.unit, rr.annotation, rr.resultErr
+}
 
-	if len(entry.Metadata) != 0 {
-		r.metadataDecoder.Reset(entry.Metadata)
-		decoded, err := r.metadataDecoder.DecodeLogMetadata()
-		if err != nil {
-			resultErr = err
+func (r *reader) readLoop() {
+	index := 0
+	for {
+		select {
+		case <-r.cancelCtx.Done():
+			for _, decoderBuf := range r.decoderBufs {
+				close(decoderBuf)
+			}
+			r.shutdownChan <- r.close()
 			return
+		default:
+			data, err := r.readChunk()
+			// TODO: Special case if err == io.EOF here?
+
+			// Distribute the decoding work in round-robin fashion so that when we
+			// read round-robin, we get the data back in the original order.
+			r.decoderBufs[index%r.numConc] <- decoderArg{
+				bytes: data,
+				err:   err,
+			}
+			index++
 		}
+	}
+}
 
-		id := r.bytesPool.Get(len(decoded.ID))
-		id.IncRef()
-		defer id.DecRef()
-		id.AppendAll(decoded.ID)
-
-		namespace := r.bytesPool.Get(len(decoded.Namespace))
-		namespace.IncRef()
-		defer namespace.DecRef()
-		namespace.AppendAll(decoded.Namespace)
-
-		r.metadataLookup[entry.Index] = Series{
-			UniqueIndex: entry.Index,
-			ID:          ts.BinaryID(id),
-			Namespace:   ts.BinaryID(namespace),
-			Shard:       uint32(decoded.Shard),
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan *readResponse) {
+	decodingOpts := r.opts.FilesystemOptions().DecodingOptions()
+	decoder := msgpack.NewDecoder(decodingOpts)
+	metadataDecoder := msgpack.NewDecoder(decodingOpts)
+
+	for arg := range inBuf {
+		readResponse := &readResponse{}
```

Comment:
I think it would be best to make responses simply a `chan readResponse` rather than a `chan *readResponse` so we can avoid having to allocate these on the heap and simply pass by value.

Reply:
Ok that makes a lot of sense actually, let me test it

Reply:
You were right about this. I didn't generate flamegraphs, but I benchmarked for throughput and changed this to pass a struct instead of a pointer didn't seem to have any affect on the readers performance.

Reply:
I'll update the code